cns/restserver/ipam.go

```diff
@@ -478,101 +627,230 @@ func (service *HTTPRestService) MarkExistingIPsAsPendingRelease(pendingIPIDs []s
+	podIPInfo := make([]cns.PodIpInfo, numDesiredIPAddresses)
+	// creating a map for the loop to check to see if the IP in the pool is one of the desired IPs
+	desiredIPMap := make(map[string]struct{})
+	// slice to keep track of IP configs to assign
+	ipConfigsToAssign := make([]cns.IPConfigurationStatus, 0)
+
 	service.Lock()
 	defer service.Unlock()
 
-	for _, ipConfig := range service.PodIPConfigState {
-		if ipConfig.IPAddress == desiredIPAddress {
-			switch ipConfig.GetState() { //nolint:exhaustive // ignoring PendingRelease case intentionally
-			case types.Assigned:
-				// This IP has already been assigned, if it is assigned to same pod, then return the same
-				// IPconfiguration
-				if ipConfig.PodInfo.Key() == podInfo.Key() {
-					logger.Printf("[AssignDesiredIPConfig]: IP Config [%+v] is already assigned to this Pod [%+v]", ipConfig, podInfo)
-				} else {
-					return podIpInfo, errors.Errorf("[AssignDesiredIPConfig] Desired IP is already assigned %+v, requested for pod %+v", ipConfig, podInfo)
-				}
-			case types.Available, types.PendingProgramming:
-				// This race can happen during restart, where CNS state is lost and thus we have lost the NC programmed version
-				// As part of reconcile, we mark IPs as Assigned which are already assigned to Pods (listed from APIServer)
-				if err := service.assignIPConfig(ipConfig, podInfo); err != nil {
-					return podIpInfo, err
+	for _, desiredIP := range desiredIPAddresses {
+		desiredIPMap[desiredIP] = struct{}{}
+	}
+
+	numIPConfigsAssigned := 0
+	for _, ipConfig := range service.PodIPConfigState { //nolint:gocritic // ignore copy
+		_, found := desiredIPMap[ipConfig.IPAddress]
+		// keep searching until the all desired IPs are found
+		if !found {
+			continue
+		}
+		switch ipConfig.GetState() { //nolint:exhaustive // ignoring PendingRelease case intentionally
+		case types.Assigned:
+			// This IP has already been assigned, if it is assigned to same pod add the IP to podIPInfo
+			if ipConfig.PodInfo.Key() == podInfo.Key() {
+				logger.Printf("[AssignDesiredIPConfigs]: IP Config [%+v] is already assigned to this Pod [%+v]", ipConfig, podInfo)
+
+				if err := service.populateIPConfigInfoUntransacted(ipConfig, &podIPInfo[numIPConfigsAssigned]); err != nil {
+					//nolint:goerr113 // return error
+					return []cns.PodIpInfo{}, fmt.Errorf("[AssignDesiredIPConfigs] Failed to assign IP %+v requested for pod %+v", ipConfig, podInfo)
 				}
-			default:
-				return podIpInfo, errors.Errorf("[AllocateDesiredIPConfig] Desired IP is not available %+v", ipConfig)
+				numIPConfigsAssigned++
+			} else {
+				//nolint:goerr113 // return error
+				return []cns.PodIpInfo{}, fmt.Errorf("[AssignDesiredIPConfigs] Desired IP is already assigned %+v, requested for pod %+v", ipConfig, podInfo)
 			}
-			err := service.populateIPConfigInfoUntransacted(ipConfig, &podIpInfo)
-			return podIpInfo, err
+		case types.Available, types.PendingProgramming:
```

Comment:
can we add a todo / question on why we would attempt to assign ip which is pendingProgramming?
If the IP state is pendingProgramming for the desired IP, we should fail the request as it may cause the data path issue if we assign it. @rbtr

Reply:
I can add something. I thought this was odd myself when I first saw it in the code. I believe the comment below is supposed to explain why and it sounds like we need this for when CNS restarts and marking everything back as assigned that has already been assigned. I would like to know more about this though.
```
// This race can happen during restart, where CNS state is lost and thus we have lost the NC programmed version
			// As part of reconcile, we mark IPs as Assigned which are already assigned to Pods (listed from APIServer)
```